src/node_osrm.cpp

```diff
@@ -4,594 +4,507 @@
 
-    if (args.Length() < 2) {
-        NanThrowTypeError("two arguments required");
-        NanReturnUndefined();
+    auto path = params->Get(Nan::New("path").ToLocalChecked());
+    auto shared_memory = params->Get(Nan::New("shared_memory").ToLocalChecked());
+    if (!path->IsUndefined())
+    {
+        lib_config->server_paths["base"] =
+            *v8::String::Utf8Value(Nan::To<v8::String>(path).ToLocalChecked());
+    }
+    if (!shared_memory->IsUndefined())
+    {
+        lib_config->use_shared_memory = Nan::To<bool>(shared_memory).FromJust();
     }
 
-    if (!args[0]->IsObject()) {
-        NanThrowTypeError("first arg must be an object");
-        NanReturnUndefined();
+    if (path->IsUndefined() && !lib_config->use_shared_memory)
+    {
+        Nan::ThrowError("shared_memory must be enabled if no path is "
+                        "specified");
+        return libosrm_config_ptr();
     }
 
-    Local<Object> obj = args[0]->ToObject();
+    return lib_config;
+}
 
-    route_parameters_ptr params = make_unique<RouteParameters>();
+boost::optional<std::vector<FixedPointCoordinate>> parseCoordinateArray(const v8::Local<v8::Array> &coordinates_array)
+{
+    boost::optional<std::vector<FixedPointCoordinate>> resulting_coordinates;
+    std::vector<FixedPointCoordinate> temp_coordinates;
+
+    for (uint32_t i = 0; i < coordinates_array->Length(); ++i)
+    {
+        v8::Local<v8::Value> coordinate = coordinates_array->Get(i);
 
-    params->zoom_level = 18; //no generalization
-    params->print_instructions = false; //turn by turn instructions
-    params->alternate_route = true; //get an alternate route, too
-    params->geometry = true; //retrieve geometry of route
-    params->compression = true; //polyline encoding
-    params->check_sum = 0; //see wiki
-    params->service = "viaroute"; //that's routing
-    params->output_format = "json";
-    params->jsonp_parameter = ""; //set for jsonp wrapping
-    params->language = ""; //unused atm
+        if (!coordinate->IsArray())
+        {
+            Nan::ThrowError("coordinates must be an array of (lat/long) pairs");
+            return resulting_coordinates;
+        }
+
+        v8::Local<v8::Array> coordinate_pair = v8::Local<v8::Array>::Cast(coordinate);
+        if (coordinate_pair->Length() != 2)
+        {
+            Nan::ThrowError("coordinates must be an array of (lat/long) pairs");
+            return resulting_coordinates;
+        }
 
-    if (!obj->Has(NanNew("coordinates"))) {
-        NanThrowError("must provide a coordinates property");
-        NanReturnUndefined();
+        temp_coordinates.emplace_back(
+            static_cast<int>(coordinate_pair->Get(0)->NumberValue() * COORDINATE_PRECISION),
+            static_cast<int>(coordinate_pair->Get(1)->NumberValue() * COORDINATE_PRECISION));
     }
+    resulting_coordinates =
+        boost::make_optional<std::vector<FixedPointCoordinate>>(std::move(temp_coordinates));
```

Comment:
There should be no need to specify the type here, if I'm not missing something.

The `make_*` functions are for type-deduction, that is instead of writing

    my_class<int, double> x{1, 4.2};

you would write

    auto x = make_my_class(1, 4.2);

The only reason for having those `make_*` functions is that there is no type deduction in calling a constructor. (There was a proposal for this at some point in time, if I recall correctly)